app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,200 @@
+package org.kiwix.kiwixmobile.wifi_hotspot;
+
+import android.app.Notification;
+import android.app.NotificationChannel;
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.app.Service;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Binder;
+import android.os.Build;
+import android.os.IBinder;
+import android.util.Log;
+import android.widget.Toast;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.core.app.NotificationCompat;
+import org.kiwix.kiwixmobile.R;
+import org.kiwix.kiwixmobile.utils.Constants;
+import org.kiwix.kiwixmobile.utils.ServerUtils;
+import org.kiwix.kiwixmobile.webserver.ZimHostCallbacks;
+import org.kiwix.kiwixmobile.webserver.WebServerHelper;
+import org.kiwix.kiwixmobile.webserver.ZimHostActivity;
+
+import static org.kiwix.kiwixmobile.webserver.ZimHostActivity.SELECTED_ZIM_PATHS_KEY;
+
+/**
+ * HotspotService is used to add a foreground service for the wifi hotspot.
+ * Created by Adeel Zafar on 07/01/2019.
+ */
+
+public class HotspotService extends Service {
+
+  public static final String ACTION_TURN_ON_AFTER_O = "Turn_on_hotspot_after_oreo";
+  public static final String ACTION_TURN_OFF_AFTER_O = "Turn_off_hotspot_after_oreo";
+  public static final String ACTION_IS_HOTSPOT_ENABLED = "Is_hotspot_enabled";
+  public static final String ACTION_START_SERVER = "start_server";
+  public static final String ACTION_STOP_SERVER = "stop_server";
+
+  private static final int HOTSPOT_NOTIFICATION_ID = 666;
+  private static final String ACTION_STOP = "hotspot_stop";
+  private static final String TAG = "HotspotService";
+  private WifiHotspotManager hotspotManager;
+  private BroadcastReceiver stopReceiver;
+  private NotificationManager notificationManager;
+  private NotificationCompat.Builder builder;
+  private ZimHostCallbacks zimHostCallbacks;
+  private final IBinder serviceBinder = new HotspotBinder();
+  private WebServerHelper webServerHelper;
+
+  @Override public void onCreate() {
+
+    super.onCreate();
+
+    hotspotManager = new WifiHotspotManager(this);
+
+    stopReceiver = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        if (intent != null && intent.getAction().equals(ACTION_STOP)) {
```

Comment:
You should put constants on the left hand side of an expression.
`intent.getAction().equals(ACTION_STOP)` could produce an NPE if getAction is null
`ACTION_STOP.equals(intent.getAction())` can not throw an NPE

Reply:
Also I requested that the broadcast receiver be removed and the intent handling put to the service in an earlier comment. It is pretty easy to lose comments because you have to load through everything in the conversation tab and perhaps you already have this queued up.

Reply:
This broadcast receiver is for the **STOP BUTTON** in the foreground notification. I'm not sure if we can handle this intent in the `onStartCommand()`.

Reply:
One way I can think of is if we have our notification in the layout. We can attach a `onClickListener` using the `RemoteViews`. But we don't have a layout for our notification.

Reply:
Have you tried just using a pending intent for HotspotService? I would be pretty surprised if it didn't work.
PendingIntents are wrappers around intents. We use them when we need to allow another program or the system to be able to act like our program. So if you were able to send the message to the service at the time you created the PendingIntent then there should be no reason it won't work.

There is infact an example of this already in the codebase in DownloadService but I should really just delete that class.